### PR TITLE
Human readable x axis for pycbc_inference_plot_samples

### PR DESCRIPTION
**Pull request by stevereyes01**
Change the x-axis for plot samples. It currently writes iterations as the label, even though it's actually plotting interval lengths.

If `thin-interval` is not specified, the code is supposed to default to plotting in intervals of ACL.

This patch does not use the ACL for each parameter (which I think it should do). Maybe that can go in a future patch. Probably a good idea would be to calculate the ACL for each parameter, and plot different x-axes in the case that these ACL differ from one another.

Current behavior with thin-interval = 1000:
https://sugwg-jobs.phy.syr.edu/~steven.reyes/O2/test_PE/params_chain.png

Current behavior with no thin-interval given:
https://sugwg-jobs.phy.syr.edu/~steven.reyes/O2/test_PE/params_chain_acl.png

Prior behavior would just write Iterations as the x-axis label but would generate all of the same ticks for both plots.

### CODE REVIEW

**Comment by soumide1102** [comment]
Actually, the `pycbc_inference_plot_samples` executable should not be given a `--thin-interval` other than `1` (ie. it should not be allowed to use the autocorrelation length from the file) because then the plot it would make would make could look crazy and would not make much sense if the interval that the user provides is too big. For eg. looking at your 1000 interval plot, it's hard to track how each of the walkers are evolving with time. So can you just make the executable raise an error if a `--thin-interval` other than 1 is provided? So, basically you do not have to change any labels. I think @cmbiwer and I had talked about this once, but never executed it.

**Comment by stevereyes01** [comment]
So you want it to plot iterations, and never intervals?

**Comment by stevereyes01** [comment]
The help message says that this code can plot `--thin-intervals` and that if it's not given it will use auto-correlation lengths as the `thin-interval`.

**Comment by soumide1102** [comment]
Also, if a `--thin-start` !=0 and `--thin-end`!=last iteration is provided is provided, the x-axis is of the plot does not show those two values as the start and the end. But it always starts from 0. I think it would be nice to it be able to plot from say 79999 to 99999 in the x-axis, if those are the `--thin-start` and `--thin-end` provided.

File: bin/inference/pycbc_inference_plot_samples [hunk 1]
```diff
@@ -78,11 +78,20 @@ for i, arg in enumerate(parameters):
...
+        axs[i].plot(x, y[arg], alpha=0.25)
 
         # y-axis label
         axs[i].set_ylabel(labels[i])
 
+# x-axis label
+if opts.thin_interval is not None:
+    axs[-1].set_xlabel("Intervals [%s iteration(s) per interval]" % opts.thin_interval)
+# Default to old behavior
+else :
```
Comment (cmbiwer) [inline bin/inference/pycbc_inference_plot_samples]: PEP8.

**Comment by cmbiwer** [comment]
I don't understand some of the conversation going on here about this executable.

> Change the x-axis for plot samples. It currently writes iterations as the label, even though it's actually plotting interval lengths.

I like the look of the x-axis on the plots. Its nice and the meaning is more clear.

> This patch does not use the ACL for each parameter (which I think it should do). Maybe that can go in a future patch. Probably a good idea would be to calculate the ACL for each parameter, and plot different x-axes in the case that these ACL differ from one another.

What do you mean here? Is there some loss of functionality? Looks like you label for ACL already.

> Actually, the pycbc_inference_plot_samples executable should not be given a --thin-interval other than 1 (ie. it should not be allowed to use the autocorrelation length from the file) because then the plot it would make would make could look crazy and would not make much sense if the interval that the user provides is too big. For eg. looking at your 1000 interval plot, it's hard to track how each of the walkers are evolving with time. So can you just make the executable raise an error if a --thin-interval other than 1 is provided? So, basically you do not have to change any labels. I think @cmbiwer and I had talked about this once, but never executed it.

Wait. I don't think I would have said that, I actually use the thinning options with this executable from time to time. Please don't remove that.

I think the conversation about this executable was to allow it to plot a hexplot/density plot so you can see the density of the walkers.

I like the label change. Please keep it.

**Comment by stevereyes01** [comment]
@cmbiwer Isn't there a different ACL for each parameter? It's just using the 1 value written to file. You guys know a lot more about how these work. I just wanted to change the x-axis label to something that makes sense to my future self.

**Comment by soumide1102** [comment]
Okay. Then I'm misremembering something I guess.

**Comment by cmbiwer** [comment]
Follow me down the rabbit hole @stevereyes01 

ACL gets written to file in ``pycbc_inference`` by a class method (``write_acls``) of the samplers here: https://github.com/ligo-cbc/pycbc/blob/master/bin/inference/pycbc_inference#L389

Can look in sampler modules to see that those functions are writing the maximum ACL:
  * emcee: https://github.com/ligo-cbc/pycbc/blob/master/pycbc/inference/sampler_emcee.py#L903

We note it calls ``super`` as well so we need to know what the super class does. The super class is in ``pycbc.inference.sampler_base``.

``sampler_kombine`` doesn't have one so we look at its super class in ``sampler_base``as well. The function to look at is here: https://github.com/ligo-cbc/pycbc/blob/master/pycbc/inference/sampler_base.py#L913

And we see that for both kombine and emcee it writes the maximum ACL over all parameters to ``fp.attrs`` but also writes the (maximum if multiple temperatures) ACL for each parameter in the corresponding HDF group eg. ``fp['samples/mass1'].attrs``.

**Comment by stevereyes01** [comment]
That's quite a bit. I'll leave it to the top level ACL, `fp.attrs` (?)

**Comment by cmbiwer** [comment]
You can leave it for now sure but we should probably switch to allowing parameter-specific ACL from the command line.

**Comment by cdcapano** [comment]
@stevereyes01 @soumide1102 @cmbiwer Since you guys are editing this code, it would be useful if this code plotted the average parameter value over all of the (same-temperature) walkers. @farr had requested this in issue #1807; it's the only remaining thing listed in that issue that hasn't been solved. It would make these plots much more readable. You could still plot individual walkers if a command-line option were provided, like `plot_acf`.

Also, I agree with @cmbiwer, don't remove the ability to use a thin interval.

**Comment by cdcapano** [comment]
@soumide1102 @stevereyes01 Since nothing has happened with this since October, I'm going to close this in a day unless you object.

File: bin/inference/pycbc_inference_plot_samples [hunk 2]
```diff
@@ -78,11 +78,20 @@ for i, arg in enumerate(parameters):
...
+    axs[-1].set_xlabel("Autocorrelation Lengths [%s iteration(s) per interval]" % fp.attrs['acl'])
```
Comment (soumide1102) [inline bin/inference/pycbc_inference_plot_samples]: Looks like this isn't going to describe what is actually being plotted when `fp.attrs['acl']` != 1. I see that in line 83, samples from all iterations are being plotted and not the ones after every `fp.attrs['acl']` interval. So I think labeling this one as just `iterations` would be right.

**Comment by soumide1102** [comment]
I think @cdcapano 's request can be addressed in another PR. So if @stevereyes01 addresses my comment above and removes the Work in progress label, this can be merged. Also, needs to be rebased.

**Comment by ahnitz** [comment]
@stevereyes01 This seems to be abandoned. If I'm wrong, please re-open this.

**Pull request closed without merging**